host/docs/uhd_cal_tx_iq_balance.1
.TH "uhd_cal_tx_iq_balance" "1" "3.7.0" "UHD" "User Commands"
.SH NAME
uhd_cal_tx_iq_balance \- Generate TX IQ Balance Calibration Table for a UHD Device
.SH DESCRIPTION
The Universal Software Radio Peripheral Hardware Drivers handle calibration
information. This application reduces TX IQ imbalance within the specified frequency
range using transmit leakage into the receive path. Remove all external hardware
from the RF antenna ports before using this application. This can take up to
several minutes with default settings.
.LP
NOTE: This application can only be used with the WBX, SBX, XCVR2450, and RFX daughterboards.
.SH OPTIONS
.TP
\fB\-\-help\fR
help message
.TP
\fB\-\-verbose\fR
enable some verbose
.TP
\fB\-\-args\fR arg
device address args [default = ""]
.TP
\fB\-\-tx_wave_freq\fR arg (=507123)
Transmit wave frequency in Hz
.TP
\fB\-\-tx_wave_ampl\fR arg (=0.69999999999999996)
Transmit wave amplitude in counts
.TP
\fB\-\-rx_offset\fR arg (=934400)
RX LO offset from the TX LO in Hz
.TP
\fB\-\-freq_start\fR arg
Frequency start in Hz (will use beginning of frequency range if unspecified)
.TP
\fB\-\-freq_stop\fR arg
Frequency stop in Hz (will use end of frequency range if unspecified)
.TP
\fB\-\-freq_step\fR arg (=7300000)
Step size for LO sweep in Hz
.TP
\fB\-\-nsamps\fR arg (=10000)
Samples per data capture
.PP
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_cal_tx_dc_offset(1) uhd_cal_rx_iq_balance(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2012-2022 Ettus Research, A National Instruments Brand
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 004.001.000.005
* ad9361
  -  Add comment re overclocking
  -  Modify set-tx-gain procedure to update gain in one go
* b200
  -  Re-sync times
  -  Move the B200 radio control core into usrp/b200/
* cal
  -  Use safe version of set_thread_priority()
* ci
  -  Device wait to redlock scope for Vivado close
  -  Add Fedora 34 and remove Fedora 32
  -  Refactor installers and add Windows support
  -  Add custom boost version support
  -  Remove documentation-only changes from pipeline runs
  -  enable batch CI
  -  Split CI and PR pipelines for mono pipeline
  -  Enable custom CXX flags, enable -Werror
  -  Let make keep building upon failure
  -  Add clang as a compiler to all Fedora and Ubuntu containers
* cmake
  -  use LooseVersion to ensure correct version comparisons
  -  Fix rfnoc-example (CMake paths)
  -  Fix issues with static builds and CMRC
  -  Replace CMAKE_{SOURCE,BINARY}_DIR with UHD_*_DIR
  -  tests: Add build-python path to PYTHONPATH
  -  Add check for libatomic linking requirement
  -  remove duplicate entry in LIBUHD_PYTHON_GEN_SOURCE
  -  Fix VS names and use relative for images
* dbsrx
  -  Fix issue with loop variable
* debs
  -  Update upload_debs script
* devtest
  -  Clarify data type in multi_usrp_test::send_waveform()
  -  Add receive stability test to B2xx devtest
  -  Add receive stability test
* dissectors
  -  Fix whitespace formatting in CMake files
  -  Fix inclusion of glib.h and Python version
* docs
  -  Several minor manual improvements
  -  Collect all RFNoC block controllers in a module in the manual
  -  Align dependencies and bump deb package versions
  -  Clarify set/get_gpio_attr() and GPIO banks
  -  Fix GPIO documentation example
  -  x410: Fix info on loading SD card images with bmaptool
  -  Improve docs for rx_streamer::recv() on overruns
  -  sync: Update page on synchronization
  -  Fix typo in ZBX Block Diagram
  -  x4xx: Document configuring eth0 static IP
  -  Remove bmaptool instructions for writing filesystems
* examples
  -  Use cmul for gain block in-tree IP example
  -  Test all variants in gain testbench
  -  Make IQ order clear in gain RFNoC block
  -  Improve txrx_loopback_to_file (late recv, Boost, timing)
  -  Show how to use in-tree Verilog header
  -  Add x400/x410 target to RFNoC example
  -  gpio: Separate bank and port arguments
  -  usrp_power_meter: fix channel indexing when reading USRP power
* fpga
  -  Revert "Add ability to get time from Radio block"
  -  Add ability to get time from Radio block
  -  rfnoc: Add RFNoC CHDR resize module
  -  rfnoc: Add CHDR management util functions
  -  lib: Clean up axi_mux
  -  rfnoc: Add labels to axi_switch generate blocks
  -  rfnoc: Add labels to chdr_mgmt_pkt_handler
  -  rfnoc: Add documentation to chdr_xb_routing_table
  -  Shorten line length for Launchpad linter
  -  x300: Update synchronizer constraint
  -  n3xx: Update synchronizer constraint
  -  lib: Update example constraint in synchronizer
  -  Update help message for setupenv.sh
  -  Remove stale references to UHD_FPGA_DIR
  -  tools: Add UHD_FPGA_DIR definition to synthesis
  -  Set default part for sim in setupenv.sh
  -  Fix Xilinx bitfile parser for Python 3
  -  rfnoc: Fix EOB loss in DUC
  -  sim: Add PkgComplex, PkgMath, and PkgRandom
  -  lib: Clean up and document lib files
  -  x400: Remove stale information in register map
  -  ci: Add testbench pipeline
* host
  -  Revert "Add ability to get time from Radio block"
  -  Add ability to get time from Radio block
  -  python: Return mb_controller with reference_internal
  -  x4xx: Implement GPIO API
  -  Add GPIO functions to MPM RPC shim
  -  gpio: Create gpio_atr_offsets to store GPIO registers
* images
  -  Update image packager script for Python 3
* lib
  -  Remove all remaining usage of boost::numeric::bounds<>
  -  transport: Mark typecast as intended
  -  transport: Initialize _hshake_args_server
  -  rfnoc: Make implicit typecasts explicit
  -  rfnoc: Change enum node_type to enum class
  -  Add various missing includes
* libusb
  -  Remove unused context variable
* mpm
  -  x4xx: update mboard_max_rev
  -  x4xx: Allow GPIO0 and GPIO1 as port names
  -  x4xx: add DIO GPIO API configuration methods
  -  mpm: x4xx: Add checks before accessing self.dio_control
  -  Fix handling of rfic_digital_loopback argument
  -  rfdc: Tear down RFDC on teardown
  -  add X410 support for 250e6 master clock rate
  -  Expose motherboard regs for debugging
* mpmd
  -  Increase UHD-side MTU cap for 10 GbE and 1 GbE
* multi_usrp_rfnoc
  -  Revert "Reduce latency of get_time_now()"
  -  Reduce latency of get_time_now()
* n320
  -  Reduce PLL lock time
* n3x0/e3x0
  -  Remove reference to "master FP-GPIO radio"
* n3xx
  -  Add support for rev 10
  -  Fix White Rabbit
* python
  -  multi_usrp: Add set_rx_spp()
  -  multi_usrp: Fix issues in send_waveform()
  -  multi_usrp: Fix issues with recv_num_samps()
  -  Fix dropped-sample calculation in benchmark_rate.py
  -  multi_usrp: Fix overloaded function definition
  -  Add new method bindings to noc_block_base
  -  rfnoc: Change reference type for noc_block_base export
* responder
  -  Fix printw function arguments
* rfnoc
  -  radio: Fix async message handling channel checks
  -  mgmt_portal: Fix order of validity checks
  -  blocks: Minor cleanup (whitespace, typos)
  -  mgmt_portal: Remove two unused variables
  -  Add vivado-path to rfnoc_image_builder
  -  ddc: Improve unit tests and documentation
  -  duc: Fix frequency range for DUC block
  -  duc: Remove stale references to CORDIC
  -  siggen: Fix direction of rotation
* rh
  -  Fix auto DC-offset correction and auto-IQ balance APIs
* tests
  -  Remove skip_dram from streaming performance test script
  -  Fix rfnoc_graph mock nodes stop-stream command
  -  Use reference type to prevent copy
  -  Fix check in link_test
  -  Fix potential resource leak
* tools
  -  Fix rfnoc dissector build
* uhd

host/utils/uhd_cal_tx_iq_balance.cpp
//
// Copyright 2010,2012,2014 Ettus Research LLC
// Copyright 2018 Ettus Research, a National Instruments Company
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include "usrp_cal_utils.hpp"
#include <uhd/utils/safe_main.hpp>
#include <boost/math/special_functions/round.hpp>
#include <boost/program_options.hpp>
#include <chrono>
#include <complex>
#include <cstdlib>
#include <ctime>
#include <functional>
#include <iostream>

namespace po = boost::program_options;

/***********************************************************************
 * Tune RX and TX routine
 **********************************************************************/
static double tune_rx_and_tx(
    uhd::usrp::multi_usrp::sptr usrp, const double tx_lo_freq, const double rx_offset)
{
    // tune the transmitter with no cordic
    uhd::tune_request_t tx_tune_req(tx_lo_freq);
    tx_tune_req.dsp_freq_policy = uhd::tune_request_t::POLICY_MANUAL;
    tx_tune_req.dsp_freq        = 0;
    usrp->set_tx_freq(tx_tune_req);

    // tune the receiver
    double rx_freq        = usrp->get_tx_freq() - rx_offset;
    double min_fe_rx_freq = usrp->get_fe_rx_freq_range().start();
    double max_fe_rx_freq = usrp->get_fe_rx_freq_range().stop();
    uhd::tune_request_t rx_tune_req(rx_freq);
    rx_tune_req.dsp_freq_policy = uhd::tune_request_t::POLICY_MANUAL;
    rx_tune_req.dsp_freq        = 0;
    if (rx_freq < min_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - min_fe_rx_freq;
    else if (rx_freq > max_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - max_fe_rx_freq;
    usrp->set_rx_freq(rx_tune_req);

    wait_for_lo_lock(usrp);

    return usrp->get_tx_freq();
}

/***********************************************************************
 * Main
 **********************************************************************/
int UHD_SAFE_MAIN(int argc, char* argv[])
{
    std::string args, subdev, serial;
    double tx_wave_freq, tx_wave_ampl, rx_offset;
    double freq_start, freq_stop, freq_step;
    size_t nsamps;
    double precision;

    po::options_description desc("Allowed options");
    // clang-format off
    desc.add_options()
        ("help", "help message")
        ("verbose", "enable some verbose")
        ("args", po::value<std::string>(&args)->default_value(""), "device address args [default = \"\"]")
        ("subdev", po::value<std::string>(&subdev), "Subdevice specification (default: first subdevice, often 'A')")
        ("tx_wave_freq", po::value<double>(&tx_wave_freq)->default_value(507.123e3), "Transmit wave frequency in Hz")
        ("tx_wave_ampl", po::value<double>(&tx_wave_ampl)->default_value(0.7), "Transmit wave amplitude")
        ("rx_offset", po::value<double>(&rx_offset)->default_value(.9344e6), "RX LO offset from the TX LO in Hz")
        ("freq_start", po::value<double>(&freq_start), "Frequency start in Hz (do not specify for default)")
        ("freq_stop", po::value<double>(&freq_stop), "Frequency stop in Hz (do not specify for default)")
        ("freq_step", po::value<double>(&freq_step)->default_value(default_freq_step), "Step size for LO sweep in Hz")
        ("nsamps", po::value<size_t>(&nsamps), "Samples per data capture")
        ("precision", po::value<double>(&precision)->default_value(default_precision), "Correction precision (default=0.0001)")
    ;
    // clang-format on

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // print the help message
    if (vm.count("help")) {
        std::cout << boost::format("USRP Generate TX IQ Balance Calibration Table %s")
                         % desc
                  << std::endl;
        std::cout << "This application measures leakage between RX and TX on a "
                     "transceiver daughterboard to self-calibrate.\n"
                     "Note: Not all daughterboards support this feature. Refer to the "
                     "UHD manual for details.\n"
                  << std::endl;
        return EXIT_FAILURE;
    }

    // Create a USRP device
    uhd::usrp::multi_usrp::sptr usrp = setup_usrp_for_cal(args, subdev, serial);

    if (not vm.count("nsamps"))
        nsamps = size_t(usrp->get_rx_rate() / default_fft_bin_size);

    // create a receive streamer
    uhd::stream_args_t stream_args("fc32"); // complex floats
    uhd::rx_streamer::sptr rx_stream = usrp->get_rx_stream(stream_args);

    // create a transmit streamer
    uhd::tx_streamer::sptr tx_stream = usrp->get_tx_stream(stream_args);

    // create a transmitter thread
    std::atomic_flag transmit = ATOMIC_FLAG_INIT;
    transmit.test_and_set();
    auto transmitter = std::thread(
        std::bind(&tx_thread, &transmit, usrp, tx_stream, tx_wave_freq, tx_wave_ampl));

    // re-usable buffer for samples
    std::vector<samp_type> buff;

    // store the results here
    std::vector<result_t> results;

    if (not vm.count("freq_start"))
        freq_start = usrp->get_fe_tx_freq_range().start();
    if (not vm.count("freq_stop"))
        freq_stop = usrp->get_fe_tx_freq_range().stop();

    // check start and stop frequencies
    if (freq_start < usrp->get_fe_tx_freq_range().start()) {
        std::cerr << "freq_start must be " << usrp->get_fe_tx_freq_range().start()
                  << " or greater for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }
    if (freq_stop > usrp->get_fe_tx_freq_range().stop()) {
        std::cerr << "freq_stop must be " << usrp->get_fe_tx_freq_range().stop()
                  << " or less for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    // check rx_offset
    double min_rx_offset =
        usrp->get_rx_freq_range().start() - usrp->get_fe_tx_freq_range().start();
    double max_rx_offset =
        usrp->get_rx_freq_range().stop() - usrp->get_fe_tx_freq_range().stop();
    if (rx_offset < min_rx_offset or rx_offset > max_rx_offset) {
        std::cerr << "rx_offset must be between " << min_rx_offset << " and "
                  << max_rx_offset << " for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    std::cout << boost::format("Calibration frequency range: %d MHz -> %d MHz")
                     % (freq_start / 1e6) % (freq_stop / 1e6)
              << std::endl;

    size_t tx_error_count = 0;
    for (double tx_lo_i = freq_start; tx_lo_i <= freq_stop; tx_lo_i += freq_step) {
        const double tx_lo = tune_rx_and_tx(usrp, tx_lo_i, rx_offset);

        // frequency constants for this tune event
        const double actual_rx_rate = usrp->get_rx_rate();
        const double actual_tx_freq = usrp->get_tx_freq();
        const double actual_rx_freq = usrp->get_rx_freq();
        const double bb_tone_freq   = actual_tx_freq + tx_wave_freq - actual_rx_freq;
        const double bb_imag_freq   = actual_tx_freq - tx_wave_freq - actual_rx_freq;

        // reset TX IQ balance
        usrp->set_tx_iq_balance(0.0);

        // set optimal RX gain setting for this frequency
        set_optimal_rx_gain(usrp, rx_stream, tx_wave_freq);

        // capture initial uncorrected value
        capture_samples(usrp, rx_stream, buff, nsamps);
        const double initial_suppression =
            compute_tone_dbrms(buff, bb_tone_freq / actual_rx_rate)
            - compute_tone_dbrms(buff, bb_imag_freq / actual_rx_rate);

        // bounds and results from searching
        double phase_corr_start = -1.0;
        double phase_corr_stop  = 1.0;
        double phase_corr_step =
            (phase_corr_stop - phase_corr_start) / (num_search_steps + 1);
        double ampl_corr_start = -1.0;
        double ampl_corr_stop  = 1.0;
        double ampl_corr_step =
            (ampl_corr_stop - ampl_corr_start) / (num_search_steps + 1);
        double best_suppression = 0;
        double best_phase_corr  = 0;
        double best_ampl_corr   = 0;
        while (phase_corr_step >= precision or ampl_corr_step >= precision) {
            for (double phase_corr = phase_corr_start + phase_corr_step;
                 phase_corr <= phase_corr_stop - phase_corr_step;
                 phase_corr += phase_corr_step) {
                for (double ampl_corr = ampl_corr_start + ampl_corr_step;
                     ampl_corr <= ampl_corr_stop - ampl_corr_step;
                     ampl_corr += ampl_corr_step) {
                    const std::complex<double> correction(ampl_corr, phase_corr);
                    usrp->set_tx_iq_balance(correction);

                    // receive some samples
                    capture_samples(usrp, rx_stream, buff, nsamps);
                    // check for TX errors in the current captured iteration
                    if (has_tx_error(tx_stream)) {
                        std::cout << "[WARNING] TX error detected! "
                                  << "Repeating current iteration" << std::endl;
                        // Undo ampl corr step
                        ampl_corr -= ampl_corr_step;
                        tx_error_count++;
                        if (tx_error_count >= MAX_NUM_TX_ERRORS) {
                            throw uhd::runtime_error(
                                "Too many TX errors. Aborting calibration.");
                        }
                        continue;
                    }
                    const double tone_dbrms =
                        compute_tone_dbrms(buff, bb_tone_freq / actual_rx_rate);
                    const double imag_dbrms =
                        compute_tone_dbrms(buff, bb_imag_freq / actual_rx_rate);
                    const double suppression = tone_dbrms - imag_dbrms;

                    if (suppression > best_suppression) {
                        best_suppression = suppression;
                        best_phase_corr  = phase_corr;
                        best_ampl_corr   = ampl_corr;
                    }
                }
            }
            phase_corr_start = best_phase_corr - phase_corr_step;
            phase_corr_stop  = best_phase_corr + phase_corr_step;
            phase_corr_step =
                (phase_corr_stop - phase_corr_start) / (num_search_steps + 1);
            ampl_corr_start = best_ampl_corr - ampl_corr_step;
            ampl_corr_stop  = best_ampl_corr + ampl_corr_step;
            ampl_corr_step  = (ampl_corr_stop - ampl_corr_start) / (num_search_steps + 1);
        }
        if (best_suppression > initial_suppression) // keep result
        {
            result_t result;
            result.freq      = tx_lo;
            result.real_corr = best_ampl_corr;
            result.imag_corr = best_phase_corr;
            result.best      = best_suppression;
            result.delta     = best_suppression - initial_suppression;
            results.push_back(result);
            if (vm.count("verbose"))
                std::cout << boost::format(
                                 "TX IQ: %f MHz: best suppression %f dB, corrected %f dB")
                                 % (tx_lo / 1e6) % result.best % result.delta
                          << std::endl;
            else
                std::cout << "." << std::flush;
        }

        // Reset underrun counts, start a new counter for the next frequency
        tx_error_count = 0;
    }
    std::cout << std::endl;

    // stop the transmitter
    transmit.clear();
    transmitter.join();

    store_results(results, "TX", "tx", "iq", serial);

    return EXIT_SUCCESS;
}
